routing/index_graph_loader.cpp

```diff
@@ -0,0 +1,107 @@
+#include "routing/index_graph_loader.hpp"
+
+#include "routing/index_graph_serialization.hpp"
+#include "routing/restriction_loader.hpp"
+#include "routing/routing_exceptions.hpp"
+
+#include "coding/file_container.hpp"
+
+#include "base/assert.hpp"
+
+namespace
+{
+using namespace routing;
+
+class IndexGraphLoaderImpl final : public IndexGraphLoader
+{
+public:
+  IndexGraphLoaderImpl(Index & index, shared_ptr<NumMwmIds> numMwmIds,
+                       shared_ptr<VehicleModelFactory> vehicleModelFactory,
+                       shared_ptr<EdgeEstimator> estimator, shared_ptr<TrafficStash> trafficStash);
+
+  // IndexGraphLoader overrides:
+  virtual IndexGraph & GetIndexGraph(NumMwmId numMwmId) override;
+
+private:
+  void Load(NumMwmId mwmId);
+
+  Index & m_index;
+  shared_ptr<NumMwmIds> m_numMwmIds;
+  shared_ptr<VehicleModelFactory> m_vehicleModelFactory;
+  shared_ptr<EdgeEstimator> m_estimator;
+  shared_ptr<TrafficStash> m_trafficStash;
+  unordered_map<NumMwmId, unique_ptr<IndexGraph>> m_graphs;
+};
+
+IndexGraphLoaderImpl::IndexGraphLoaderImpl(Index & index, shared_ptr<NumMwmIds> numMwmIds,
+                                           shared_ptr<VehicleModelFactory> vehicleModelFactory,
+                                           shared_ptr<EdgeEstimator> estimator,
+                                           shared_ptr<TrafficStash> trafficStash)
+  : m_index(index)
+  , m_numMwmIds(numMwmIds)
+  , m_vehicleModelFactory(vehicleModelFactory)
+  , m_estimator(estimator)
+  , m_trafficStash(trafficStash)
+{
+  CHECK(m_numMwmIds, ());
+  CHECK(m_vehicleModelFactory, ());
+  CHECK(m_estimator, ());
+  CHECK(m_trafficStash, ());
+}
+
+IndexGraph & IndexGraphLoaderImpl::GetIndexGraph(NumMwmId numMwmId)
+{
+  auto it = m_graphs.find(numMwmId);
+  if (it != m_graphs.end())
+    return *it->second;
+
+  Load(numMwmId);
+
+  it = m_graphs.find(numMwmId);
+  CHECK(it != m_graphs.end(), ());
+  return *it->second;
+}
+
+void IndexGraphLoaderImpl::Load(NumMwmId numMwmId)
+{
+  platform::CountryFile const & file = m_numMwmIds->GetFile(numMwmId);
+  MwmSet::MwmHandle handle = m_index.GetMwmHandleByCountryFile(file);
+  if (!handle.IsAlive())
```

Comment:
Is it actually an error? I mean, during route construction some mwms may be absent, and route algorithm should just ignore these regions.